app/models/foreman_snapshot_management/snapshot.rb

```diff
@@ -4,70 +4,125 @@ class Snapshot
     include ActiveModel::Conversion
     include ActiveModel::Model
     include ActiveModel::Dirty
+    include ActiveModel::ForbiddenAttributesProtection
 
     define_model_callbacks :create, :save, :destroy, :revert
-    attr_accessor :id, :vmware_snapshot, :name, :description, :host_id
+    attr_accessor :id, :raw_snapshot, :name, :description, :host_id, :parent
 
-    def self.add_snapshot_with_children(snapshots, host, vmware_snapshot)
-      snapshots[vmware_snapshot.ref] = new_from_vmware(host, vmware_snapshot)
-      vmware_snapshot.child_snapshots.each do |snap|
-        add_snapshot_with_children(snapshots, host, snap)
-      end
+    def self.root_for_host(host)
+      raw_snapshot = host.compute_resource.get_snapshots(host.uuid).first
+      new_from_vmware(host, raw_snapshot)
```

Comment:
There is a problem with these lines. On a host without a snapshot, `first` returns `nil`. `new_from_vmware` cannot handle this. At the moment i do not see, whether it's safe to just return `nil` immediately.
I guess, `all_for_host` must return an empty list in that case.
I also infer, you are using the plugin extensively, so none of your hosts is left without snapshots. ;)